onnxruntime/core/util/thread_utils.cc

```diff
@@ -13,55 +13,81 @@ static inline std::vector<size_t> GenerateVectorOfN(size_t n) {
   std::vector<size_t> ret(n);
   std::iota(ret.begin(), ret.end(), 0);
   return ret;
-                }
+}
+
+struct NumCoresUtil {
 #ifdef _WIN32
-		// This function doesn't support systems with more than 64 logical processors
-		static std::vector<size_t> GetNumCpuCores() {
-			// Indeed 64 should be enough. However, it's harmless to have a little more.
-			SYSTEM_LOGICAL_PROCESSOR_INFORMATION buffer[256];
-			DWORD returnLength = sizeof(buffer);
-			if (GetLogicalProcessorInformation(buffer, &returnLength) == FALSE) {
-                          return GenerateVectorOfN(std::thread::hardware_concurrency());
-                        }
-                        std::vector<size_t> ret;
-			int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
-			for (int i = 0; i != count; ++i) {
-				if (buffer[i].Relationship == RelationProcessorCore) {
-					ret.push_back(buffer[i].ProcessorMask);
-				}
-			}
-			if (ret.empty())
-                          return GenerateVectorOfN(std::thread::hardware_concurrency());
-                        return ret;
-                }
+  // This function doesn't support systems with more than 64 logical processors
+  static std::vector<size_t> GetNumCpuCores() {
+    // Indeed 64 should be enough. However, it's harmless to have a little more.
+    SYSTEM_LOGICAL_PROCESSOR_INFORMATION buffer[256];
+    DWORD returnLength = sizeof(buffer);
+    if (GetLogicalProcessorInformation(buffer, &returnLength) == FALSE) {
+      return GenerateVectorOfN(std::thread::hardware_concurrency());
+    }
+    std::vector<size_t> ret;
+    int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
+    for (int i = 0; i != count; ++i) {
+      if (buffer[i].Relationship == RelationProcessorCore) {
+        ret.push_back(buffer[i].ProcessorMask);
+      }
+    }
+    if (ret.empty())
+      return GenerateVectorOfN(std::thread::hardware_concurrency());
+    return ret;
+  }
 #else
-		static std::vector<size_t> GetNumCpuCores() {
-			return GenerateVectorOfN(std::thread::hardware_concurrency() / 2);
-		}
+  static std::vector<size_t> GetNumCpuCores() {
+    return GenerateVectorOfN(std::thread::hardware_concurrency() / 2);
```

Comment:
Any reason to use only half of concurrency for other OS?

Reply:
Because there isn't a system API to get the number of cpu cores on Linux. We have to use a third-party lib like [hwloc](https://www.open-mpi.org/projects/hwloc/)

Reply:
For example, given a system with 4 cores, hardware_concurrency() is 4. Here we return a vector of size 2. That means we only use the first 2 cores. It limits the full potential of thread pool.

Reply:
You're right.  The code were there for a long long time, just nobody tried to change it.